### PR TITLE
Decouple timers from system clock

### PR DESCRIPTION
**Pull request by carlescufi**
Second split out from #16469. Includes #16957.
Do not merge until #16957 is merged

### CODE REVIEW

**Comment by carlescufi** [comment]
Missing files to be updated by @pizi-nordic:

```
soc/arm/silabs_exx32/efr32mg12p/dts_fixup.h
soc/arm/silabs_exx32/efr32fg1p/dts_fixup.h
```

**Review by pizi-nordic — APPROVED** [review]

**Comment by galak** [comment]
* This needs review for cases that the DT cpu freq is being used [NXP USB is one example]
* Scrub any uses of CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC that should possibly be sys_clock_hw_cycles_per_sec()

File: drivers/adc/adc_stm32.c [hunk 1]
```diff
@@ -557,7 +557,7 @@ static int adc_stm32_init(struct device *dev)
 		LOG_ERR("ADC clock rate get error.");
 	}
 
-	wait_cycles = CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC / adc_rate *
+	wait_cycles = SystemCoreClock / adc_rate *
```
Comment (galak) [inline drivers/adc/adc_stm32.c]: Why is this `SystemCoreClock` vs sys_clock_hw_cycles_per_sec()?

**Review by galak — CHANGES_REQUESTED** [review]

File: drivers/adc/adc_stm32.c (hunk 1 above)
Comment (pizi-nordic) [reply drivers/adc/adc_stm32.c]: Requested by code owner for this file some review rounds ago. Note, that there should be CPU frequency as the `wait_cycles` is used to spin around:
```
	for (int i = wait_cycles; i >= 0; i--) {
	}
```

**Comment by pizi-nordic** [comment]
@nordic-krch: Could you please check if changes in subsys/logging/log_core.c are OK for you?

**Comment by galak** [comment]
@erwango drivers/adc/adc_stm32.c why use SystemCoreClock?

File: dts/bindings/usb/nxp,kinetis-usbd.yaml [hunk 2]
```diff
@@ -22,3 +22,9 @@ properties:
 
     interrupts:
       category: required
+
+    clocks:
+      type: array
+      category: required
+      description: Clock gate control information
+      generation: structures
```
[the remark this answers is not in the document]
Comment (pizi-nordic) [reply dts/bindings/usb/nxp,kinetis-usbd.yaml]: Fixed

File: dts/bindings/watchdog/arm,cmsdk-watchdog.yaml [hunk 3]
```diff
@@ -14,3 +14,9 @@ properties:
 
     reg:
       category: required
+
+    clocks:
+      type: array
+      category: required
+      description: Clock gate control information
+      generation: structures
```
[the remark this answers is not in the document]
Comment (pizi-nordic) [reply dts/bindings/watchdog/arm,cmsdk-watchdog.yaml]: Fixed

**Review by nordic-krch — APPROVED** [review]

**Comment by erwango** [comment]
@galak 

> @erwango drivers/adc/adc_stm32.c why use SystemCoreClock?

We use `SystemCoreClock` as on STM devices clock freq changes between boot time and run time. SystemCoreClock is updated  to hold the current frequency. Note that frequency change happens only at clock_control driver init:
- Default boot time frequency use to be a default low speed freq (`SystemCoreClock` is set in soc.c)
- High speed freq is set at clock_control driver init when configuring RCC IP (PLL configuration for instance). `SystemCoreClock` is then updated to the configured freq.

**Review by erwango — APPROVED** [review]

File: drivers/adc/adc_stm32.c (hunk 1 above)
Comment (galak) [reply drivers/adc/adc_stm32.c]: @erwango does this need to be `SystemCoreClock`?

**Comment by galak** [comment]
> We use `SystemCoreClock` as on STM devices clock freq changes between boot time and run time. SystemCoreClock is updated to hold the current frequency. Note that frequency change happens only at clock_control driver init:
> 
> * Default boot time frequency use to be a default low speed freq (`SystemCoreClock` is set in soc.c)
> * High speed freq is set at clock_control driver init when configuring RCC IP (PLL configuration for instance). `SystemCoreClock` is then updated to the configured freq.

So does that mean `sys_clock_hw_cycles_per_sec` will report the wrong thing?

**Comment by erwango** [comment]
@galak 

> So does that mean sys_clock_hw_cycles_per_sec will report the wrong thing?

This is indeed the case before clock_control driver init (done before kernel init)

**Comment by erwango** [comment]
@galak 

Anyway, point is for now STM32 don't use `CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC`, directly or indirectly (since `sys_clock_hw_cycles_per_sec` is still using `CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC` value), so we're in line with current PR.

I agree STM32 should also move to clock_control configuration to DT, and eventually get rid of `SystemCoreClock` and use `sys_clock_hw_cycles_per_sec` instead (once it will return dt value and/or able to return run time updated frequency). But I don't think this should block current PR.

**Review by galak — APPROVED** [review]

**Comment by galak** [comment]
Do any of these still need cleaning up:

```
drivers/clock_control/clock_stm32_ll_common.c:	 * SystemCoreClock is preferred to CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC
drivers/clock_control/clock_stm32_ll_h7.c:	 * SystemCoreClock is preferred to CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC
drivers/clock_control/clock_stm32_ll_h7.c:	SysTick_Config(CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC / 1000);
drivers/clock_control/clock_stm32_ll_h7.c:	SystemCoreClock = CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC;
drivers/timer/Kconfig:	  TSC frequency to CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC (the local APIC
drivers/timer/Kconfig:	  value of CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC will be used instead;
drivers/timer/apic_timer.c: * CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=<hz> must contain the frequency seen
drivers/timer/apic_timer.c: *     the TSC frequency to CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC. This can be
drivers/timer/apic_timer.c:	(CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
drivers/timer/rv32m1_lptmr_timer.c:    (MHZ(8) != CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC)
include/sys_clock.h:	return CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC;
include/sys_clock.h:	(CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC == 0)
include/sys_clock.h:	(CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC % CONFIG_SYS_CLOCK_TICKS_PER_SEC)
kernel/timeout.c:int z_clock_hw_cycles_per_sec = CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC;
subsys/testsuite/ztest/include/ztest.h:#define CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC 10000000
tests/benchmarks/boot_time/testcase.yaml:    filter: CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC >= 1000000
```

**Comment by carlescufi** [comment]
> Do any of these still need cleaning up:
> 
> ```
> drivers/clock_control/clock_stm32_ll_common.c:	 * SystemCoreClock is preferred to CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC
> drivers/clock_control/clock_stm32_ll_h7.c:	 * SystemCoreClock is preferred to CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC
> drivers/clock_control/clock_stm32_ll_h7.c:	SysTick_Config(CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC / 1000);
> drivers/clock_control/clock_stm32_ll_h7.c:	SystemCoreClock = CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC;
> drivers/timer/Kconfig:	  TSC frequency to CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC (the local APIC
> drivers/timer/Kconfig:	  value of CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC will be used instead;
> drivers/timer/apic_timer.c: * CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=<hz> must contain the frequency seen
> drivers/timer/apic_timer.c: *     the TSC frequency to CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC. This can be
> drivers/timer/apic_timer.c:	(CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
> drivers/timer/rv32m1_lptmr_timer.c:    (MHZ(8) != CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC)
> include/sys_clock.h:	return CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC;
> include/sys_clock.h:	(CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC == 0)
> include/sys_clock.h:	(CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC % CONFIG_SYS_CLOCK_TICKS_PER_SEC)
> kernel/timeout.c:int z_clock_hw_cycles_per_sec = CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC;
> subsys/testsuite/ztest/include/ztest.h:#define CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC 10000000
> tests/benchmarks/boot_time/testcase.yaml:    filter: CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC >= 1000000
> ```

From what I can tell, no, but we need to confirm it. Since Piotr is away what I will do is merge this PR to avoid it getting stale, and make sure he takes a look at the list once he's back.

**Comment by pizi-nordic** [comment]
```
./drivers/clock_control/clock_stm32_ll_common.c:         * SystemCoreClock is preferred to CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC
./drivers/clock_control/clock_stm32_ll_h7.c:     * SystemCoreClock is preferred to CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC
./drivers/clock_control/clock_stm32_ll_h7.c:    SysTick_Config(CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC / 1000);
./drivers/clock_control/clock_stm32_ll_h7.c:    SystemCoreClock = CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC;
```
The code above configures the clocking tree. Ideally, the information to perform this task should be taken from DTS, not the CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC. On the other hand, it also configures SysTick, which has to be set basing on CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC. I do not know STM SoC enough to say if the two things mentioned above could be easily decoupled.

```
./drivers/timer/Kconfig:          TSC frequency to CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC (the local APIC
./drivers/timer/Kconfig:          value of CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC will be used instead;
./drivers/timer/apic_timer.c: * CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=<hz> must contain the frequency seen
./drivers/timer/apic_timer.c: *     the TSC frequency to CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC. This can be
./drivers/timer/apic_timer.c:   (CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
./drivers/timer/mchp_xec_rtos_timer.c:BUILD_ASSERT_MSG(CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC == 32768,
./drivers/timer/mchp_xec_rtos_timer.c: * CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=<hz> must be set to 32768.
./drivers/timer/mchp_xec_rtos_timer.c:  (CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
./drivers/timer/mchp_xec_rtos_timer.c:#define CPT1000 ((CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC * 1000UL) \
./drivers/timer/rv32m1_lptmr_timer.c:    (MHZ(8) != CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC)

./include/sys_clock.h:  return CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC;
./include/sys_clock.h:  (CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC == 0)
./include/sys_clock.h:  (CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC % CONFIG_SYS_CLOCK_TICKS_PER_SEC)

./kernel/timeout.c:int z_clock_hw_cycles_per_sec = CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC;
```
These are parts of system timer implementation. The value of CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC has to be known there in order to configure hardware for given frequency.

```
./subsys/testsuite/ztest/include/ztest.h:#define CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC 10000000
```
Here I do not know why the value of CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC is used. Maybe someone knowing ztest will tell us why this is not taken form the board configuration.

```
./tests/benchmarks/boot_time/testcase.yaml:    filter: CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC >= 1000000
```
This looks like hack ensuring precise measurement of time. I am not sure it this is still needed after @andyross fixes in the area.

**Pull request merged**